File: server/app/services/fixed_scale.py
from decimal import Decimal, localcontext

from ..models.domain import DomainError, instant_ms


def fixed_scale_map(domain, intervals, map_id, serialize):
    if not isinstance(intervals, list) or len(intervals) > 32:
        raise DomainError("invalid_scale", "At most 32 magnified intervals are supported.")
    start, end = instant_ms(domain["from"]), instant_ms(domain["to"])
    spans = []
    for interval in intervals:
        if not isinstance(interval, dict) or set(interval) != {"from", "to", "ratio"}:
            raise DomainError("invalid_scale", "Invalid magnified interval.")
        a, b, ratio = instant_ms(interval["from"]), instant_ms(interval["to"]), interval["ratio"]
        if a >= b or type(ratio) not in (int, float) or not 1 <= ratio <= 10000:
            raise DomainError("invalid_scale", "Invalid magnified interval bounds or ratio.")
        spans.append((a, b, ratio))
    boundaries = sorted({start, end, *(t for span in spans for t in span[:2] if start < t < end)})
    with localcontext() as context:
        context.prec = 50
        masses = [Decimal(boundaries[i + 1] - a) * Decimal(str(max([1, *(ratio for left, right, ratio in spans if left <= a < right)]))) for i, a in enumerate(boundaries[:-1])]
        total, cumulative, knots = sum(masses), Decimal(0), []
        for time, mass in zip(boundaries, masses):
            knots.append({"timeMs": time, "u": serialize(cumulative / total)})
            cumulative += mass
        knots.append({"timeMs": end, "u": "1"})
    return {"mapId": map_id, "domain": domain, "knots": knots, "mode": "fixed" if intervals else "uniform", "ratio": max([1, *(span[2] for span in spans)])}

File: client/src/timeline/fixed-scale.js
import { ViewDecimal as D, decimalString, toMs } from './time-scale.js';

export function fixedScaleMap(domain, intervals = [], mapId = 'fixed-reference') {
  if (!Array.isArray(intervals) || intervals.length > 32) throw new RangeError('At most 32 magnified intervals are supported');
  const from = toMs(domain.from), to = toMs(domain.to);
  if (from >= to) throw new RangeError('Scale domain must be positive');
  const spans = intervals.map(interval => {
    if (!interval || Object.keys(interval).some(key => !['from', 'to', 'ratio'].includes(key))) throw new RangeError('Invalid magnified interval');
    const a = toMs(interval.from), b = toMs(interval.to), ratio = interval.ratio;
    if (a >= b || !Number.isFinite(ratio) || ratio < 1 || ratio > 10000) throw new RangeError('Invalid magnified interval bounds or ratio');
    return { a, b, ratio };
  });
  const boundaries = [...new Set([from, to, ...spans.flatMap(s => [s.a, s.b]).filter(t => t > from && t < to)])].sort((a, b) => a - b);
  const mass = boundaries.slice(0, -1).map((start, i) => new D(boundaries[i + 1] - start).mul(Math.max(1, ...spans.filter(s => s.a <= start && s.b > start).map(s => s.ratio))));
  const total = mass.reduce((sum, n) => sum.plus(n), new D(0));
  let cumulative = new D(0);
  const knots = boundaries.map((timeMs, i) => {
    const u = i === mass.length ? '1' : decimalString(cumulative.div(total));
    if (i < mass.length) cumulative = cumulative.plus(mass[i]);
    return { timeMs, u };
  });
  return { mapId, domain, knots, mode: intervals.length ? 'fixed' : 'uniform', ratio: Math.max(1, ...spans.map(s => s.ratio)) };
}
